_posts/2023-04-19-making-verification-compelling-visual-verification-feedback-for-dafny.markdown

```diff
@@ -0,0 +1,253 @@
+<div style="text-align:center;font-style:italic">
+Fig 2. Dynamic icons in chronological order.
+The user first opens a file and goes through A, B and C. The user makes a typo in D on line 10. Then, the user fixes the typo and the error on line 11,
+and goes in E. If verification is triggered, F, G and H happen next.
+B, F and G have 2-frame animations where the zig-zag is animated every 1/5th of a second. H is where the green gutter feels, as reported by two users, "ridiculously good".
+</div>
+
+I identified three dynamic phases for Dafny to provide feedback about its verification status. To stay true to the past-awareness principle, I slightly modified the static icons presented in the [section about static icons](#static-icons) to convey the notion of Dafny "working on verification" while still displaying previous icons.
+I present how these phases play together in [Figure 2](#fig_dynamic_gutter_icons).
+
+* **The "verifying" phase.** In this phase, illustrated in screenshots C, F and G, the icons resemble the static icons, except that they are traversed by an animated zig-zag line that indicates that Dafny is working on that code portion, but the previous results are still visible.
+* **The "obsolete" or "stale" phase.** This phase uses static and dimmed versions of the icons in the "verifying" phase. This phase is visible in screenshots A and E of [Figure 2](#fig_dynamic_gutter_icons), Dafny recognizes that something is new or has changed and that it is
+    simply parsing and resolving the new code, but hasn't started verification yet.
+* **The "unresolved" phase.** This phase uses gray scale versions of the "obsolete" icons, except for parsing and resolution errors which are indicated by red triangles containing an exclamation point, like D of [Figure 2](#fig_dynamic_gutter_icons).
+
+
+I had to solve many challenges to ensure these icons provide the most adequate feedback.
+For example, as soon as some input is entered, before parsing even occurs,
+Dafny already migrates those icons according to string edits.
+I also found the lightning metaphor to be adequate to provide the sense of "verification" on the side of the vision field without hiding the previous verification status.
+However, I observed that if we just took the rectangle for error and put a lightning on it (yellow or transparent), it would just not be visible to the side of the vision field.
+This is especially more important given that I modified Dafny to give per-method feedback.
+This is why the red rectangle is split in screenshots E and F of [Figure 2](#fig_dynamic_gutter_icons), as if the error was about to "go away".
+
+Among all the dynamic icons, note that the topmost part of [Figure 2](#fig_dynamic_gutter_icons) is less distinct than for code in error context. Without more information about the new code, I preferred the dynamic icons to reflect a kind of default hypothesis that previously verified code will probably also be verified.
```

Comment:
I'm not sure what you mean here by "the topmost part" - can you specify which line(s)?